### ISSUE TITLE
Why replace _ in url?

### ISSUE THREAD

**Issue by ushenkodmitry** [AngleSharp/AngleSharp#445]
Hi, @FlorianRappl 
Why are you replacing underscores in the url with nothing?

``` c#
    [TestMethod]
    public void TestMethod2()
    {
        //
        Url url = new Url("https://will_be_replaced.but.why/");
            //
        Assert.AreEqual("https://will_be_replaced.but.why/", url.Href);
    }
```

gives

`Assert.AreEqual failed. Expected:<https://will_be_replaced.but.why/>. Actual:<https://willbereplaced.but.why/>.`

**Comment by mausworks** [comment]
While this might be unexpected behavior &mdash; it is according to spec on valid hostnames. RFC 1124 and RFC 952 mandates that host names may not contain underscores, and as far a I know, there aren't any valid domain names that do. That being said; one would expect an exception to be raised in this particular case. Simply stripping underscores from the hostname borders on _really_ defensive programming.

From Wikipedia:

> The Internet standards for protocols mandate that component hostname labels may contain only the ASCII letters 'a' through 'z' (in a case-insensitive manner), the digits '0' through '9', and the hyphen ('-'). The original specification of hostnames in RFC 952, mandated that labels could not start with a digit or with a hyphen, and must not end with a hyphen. However, a subsequent specification (RFC 1123) permitted hostname labels to start with digits. **No other symbols, punctuation characters, or white space are permitted.**

**Comment by FlorianRappl** [comment]
The URL parser _never_ throws an exception. This is by design. Your browser also does not throw an exception. If you really want to know that you provided a valid or invalid URL (i.e., that the parser modified the URL) check the `IsInvalid` property.

All this being said: The URL parser follows the algorithms from the WHATWG group. What you see is the behavior that is also seen from standard web browsers.

**Comment by ushenkodmitry** [comment]
> What you see is the behavior that is also seen from standard web browsers

What behavior do you mean? I can reach [https://loony_picture.dirty.ru/](url) using any browser, but I can not reach it using BrowsingContext, becouse it uses Url that converts the url above to [https://loonypicture.dirty.ru/](url) and targets page that does not exist. 
Another example is [https://black_white.dirty.ru/](url) and [https://blackwhite.dirty.ru/](url) -- both are absolutely different resources.

> All this being said: The URL parser follows the algorithms from the WHATWG group.

Url class annotations states that its implemented according to RFC3986. But on pages 12-13 of this specification we can read:

> 2.3.  Unreserved Characters
> 
> Characters that are allowed in a URI but do not have a reserved
>    purpose are called unreserved.  These include uppercase and lowercase
>    letters, decimal digits, hyphen, period, underscore, and tilde.
> 
> ```
>   unreserved  = ALPHA / DIGIT / "-" / "." / "_" / "~"
> ```
> 
> Berners-Lee, et al.         Standards Track                    [Page 13]
> 
> RFC 3986                   URI Generic Syntax               January 2005
> 
>    URIs that differ in the replacement of an unreserved character with
>    its corresponding percent-encoded US-ASCII octet are equivalent: they
>    identify the same resource.  However, URI comparison implementations
>    do not always perform normalization prior to comparison (see Section
>    6).  For consistency, percent-encoded octets in the ranges of ALPHA
>    (%41-%5A and %61-%7A), DIGIT (%30-%39), hyphen (%2D), period (%2E),
>    underscore (%5F), or tilde (%7E) should not be created by URI
>    producers and, when found in a URI, should be decoded to their
>    corresponding unreserved characters by URI normalizers.

So, it seem that underscore is allowed character. This just an excerpt from the specification, but I bet that the whole document says nothing about underscore replacement.

@mausworks paying attention to all above -- yes, this is absolutely unexpected behavior for me. BrowsingContext loads absolutely different address.

**Comment by mausworks** [comment]
@ushenkodmitry while it is true that underscores are allowed in URL:s (such as in the path, query or fragment), they are not allowed in the hostname part, as both I and @FlorianRappl described. 

Regardless; I think `dirty.ru` should learn conform to established web standard, not the other way around. Using underscores in hostnames _is_ invalid according to both RFC 1124 and RFC 952. Maybe the coders of that website should read up them.

**Comment by FlorianRappl** [comment]
@ushenkodmitry see [WHATWG URL specification](https://url.spec.whatwg.org). Also Google has a nice paragraph about why URLs with underscores will be downrated.

Also please read the [Wikipedia entry on host names](https://en.wikipedia.org/wiki/Hostname):

> While a hostname may not contain other characters, such as the underscore character (_), other DNS names may contain the underscore.[4] Systems such as DomainKeys and service records use the underscore as a means to assure that their special character is not confused with hostnames. For example, _http._sctp.www.example.com specifies a service pointer for an SCTP capable webserver host (www) in the domain example.com. Note that some applications (e.g. Microsoft Internet Explorer) won't work correctly if any part of the hostname contains an underscore character.[5]

**Comment by ushenkodmitry** [comment]
Well, I understand yours point, but I do not accept )). The resource is reachable with any web browser (checked just now). It is reachable with HttpClient and all web APIs within .Net. Why it can not be reachable by BrowsingContext? 
Whould you use a browser which renders https://blackwhite.dirty.ru/ to you, even you requesting https://black_white.dirty.ru/? The problem of cource not just in dirty.ru. I'm sure there are other.

**Comment by FlorianRappl** [comment]
Agreed - I think we need to tolerate this. Fixed as PR #446.
